src/random.c

```diff
@@ -4,13 +4,89 @@
 #include <alloca.h>
 #endif
 
+// IWRAM common
+rng_value_t gRngValue;
+rng_value_t gRng2Value;
+
+#if HQ_RANDOM == TRUE
+EWRAM_DATA static volatile u8 sUnknown = 1;
+#define RNG_LOOP_LOCK sUnknown
+
+
+static void SFC32_Seed(struct Sfc32State *state, u16 seed)
+{
+    u32 i;
+    state->a = state->b = 0;
+    state->c = seed;
+    state->ctr = 1;
+    for(i = 0; i < 16; i++)
+    {
+        _SFC32_Next(state);
+    }
+}
+
+u32 NAKED Random32(void)
+{
+    asm(".thumb\n\
+    push {r4, r5, r6}\n\
+    mov r6, #11\n\
+    ldr r5, =gRngValue\n\
+    ldmia r5!, {r1, r2, r3, r4}\n\
+    @ e (result) = a + b + d++\n\
+    add r1, r1, r2\n\
+    add r0, r1, r4\n\
+    add r4, r4, #1\n\
+    @ a = b ^ (b >> 9)\n\
+    lsr r1, r2, #9\n\
+    eor r1, r1, r2\n\
+    @ b = c + (c << 3) [c * 9]\n\
+    lsl r2, r3, #3\n\
+    add r2, r2, r3\n\
+    @ c = rol(c, 21) + e\n\
+    ror r3, r3, r6\n\
+    add r3, r3, r0\n\
+    sub r5, r5, #16\n\
+    stmia r5!, {r1, r2, r3, r4}\n\
+    pop {r4, r5, r6}\n\
+    bx lr\n\
+    .ltorg"
+    );
+}
+
+u32 Random2_32(void)
+{
+    return _SFC32_Next(&gRng2Value);
+}
+
+void SeedRng(u16 seed)
+{
+    SFC32_Seed(&gRngValue, seed);
+    RNG_LOOP_LOCK = 0;
+}
+
+void SeedRng2(u16 seed)
+{
+    SFC32_Seed(&gRng2Value, seed);
+}
+
+void AdvanceRandom(void)
+{
+    if (RNG_LOOP_LOCK == 0)
+        Random32();
+}
+
+#define LOOP_RANDOM_START \
+    struct Sfc32State *const state = &gRngValue; \
+    RNG_LOOP_LOCK = 1
```

Comment:
Also now that i think about it, could you explain why we need to lock the RNG? Is it that an interrupt could occur while `Random32` is updating the state, thus causing it to contain a half-updated state during the interrupt handler?

Reply:
Yes, however this can’t happen with the ASM implementation of `Random32()` because it uses ldm/stm, so it doesn’t lock. In practice, GCC tends to not use ldm/stm, so the lock is needed there.

Reply:
Also, I know it’s the idiomatic thing to do, but why doesn’t `SHUFFLE_IMPL` do it?

Reply:
Ohhhh, sorry, I'm an idiot who shouldn't review things just after waking up! You can't use `do-while` because you're declaring a variable you want available in `LOOP_RANDOM`.

I don't think there's any particular reason `SHUFFLE_IMPL` couldn't work that way. I suppose it just didn't because you wouldn't expect somebody to write `if (...) SHUFFLE_IMPL` (which isn't a particularly convincing argument when it could have done so, and harmlessly).

Reply:
> Yes, however this can’t happen with the ASM implementation of `Random32()` because it uses ldm/stm, so it doesn’t lock. In practice, GCC tends to not use ldm/stm, so the lock is needed there.

Oh, interesting! Do you have a source for `ldm`/`stm` being atomic wrt interrupts? I tried looking for one this morning, but I think the best I found implied the opposite (possibly for a later ARM version than the ARMv4).

Reply:
Section 2.9.1 of the ARM7TDMI reference manual states that the processor cannot interrupt LDM/STM instructions as an aside. To me, this makes sense, because the processor has nowhere to store the progress done by those instructions.

Reply:
Awesome, thanks! I'd expected it to be atomic for the same reasons, but the SO answer I was looking at seemed to think maybe it just replays the instruction (which I think would provided the write-back occurs last, and you don't load to the register being written-back into).